pkg/querier/http.go

```diff
@@ -58,46 +62,62 @@ func directionParam(values url.Values, name string, def logproto.Direction) (log
 	return logproto.Direction(d), nil
 }
 
-// QueryHandler is a http.HandlerFunc for queries.
-func (q *Querier) QueryHandler(w http.ResponseWriter, r *http.Request) {
-	params := r.URL.Query()
-	query := params.Get("query")
+func httpRequestToQueryRequest(httpRequest *http.Request) (*logproto.QueryRequest, error) {
+	params := httpRequest.URL.Query()
+	now := time.Now()
+	queryRequest := logproto.QueryRequest{
+		Regex: params.Get("regexp"),
+		Query: params.Get("query"),
+	}
+
 	limit, err := intParam(params, "limit", defaultQueryLimit)
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
 	}
+	queryRequest.Limit = uint32(limit)
 
-	now := time.Now()
-	start, err := unixNanoTimeParam(params, "start", now.Add(-defaulSince))
+	queryRequest.Start, err = unixNanoTimeParam(params, "start", now.Add(-defaulSince))
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
 	}
 
-	end, err := unixNanoTimeParam(params, "end", now)
+	queryRequest.End, err = unixNanoTimeParam(params, "end", now)
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
 	}
 
-	direction, err := directionParam(params, "direction", logproto.BACKWARD)
+	queryRequest.Direction, err = directionParam(params, "direction", logproto.BACKWARD)
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
 	}
 
-	request := logproto.QueryRequest{
-		Query:     query,
-		Limit:     uint32(limit),
-		Start:     start,
-		End:       end,
-		Direction: direction,
-		Regex:     params.Get("regexp"),
+	return &queryRequest, nil
+}
+
+func writeHTTPErrorResponse(err error, defaultStatusCode int, w http.ResponseWriter) { // nolint
+	statusCode := defaultStatusCode
+	var errMessage string
+
+	if httpResponse, isHTTPResponse := httpgrpc.HTTPResponseFromError(err); isHTTPResponse {
+		errMessage = string(httpResponse.Body)
+		statusCode = int(httpResponse.Code)
+	} else {
+		errMessage = err.Error()
+	}
+
+	http.Error(w, errMessage, statusCode)
+}
```

Comment:
You should be able to use this: https://github.com/weaveworks/common/blob/master/httpgrpc/server/server.go#L169

Reply:
@tomwilkie Pushed this change